### PR TITLE
Add ability to specify domains, lists, clients and group names as arrays

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Add ability to specify domains, lists, clients and group names as arrays

Screenshot from 2023-09-06 18-47-17
Screenshot from 2023-09-06 18-47-23
Screenshot from 2023-09-06 18-47-27

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by DL6ER** [comment]
Changing this to "ready for review" and kindly asking for the same

**Comment by DL6ER** [comment]
This branch continues to work with the current  `development-v6` web branch. However, if you specifically want to review the core of this PR (being able to add multiple at once), you'll need the web interface branch `new/antigravity` which contains the necessary changes for this PR (sending multiple domains as an array of strings).

**Comment by DL6ER** [comment]
Rebased on latest `development-v6`

**Comment by DL6ER** [comment]
The spellcheck issue will be fixed in https://github.com/pi-hole/FTL/pull/1649 by https://github.com/pi-hole/FTL/commit/28358d45aa597f489c9bc054ea03b71638017c11

**Comment by DL6ER** [comment]
This PR has recently been affected by a bug solved in https://github.com/pi-hole/FTL/pull/1659 and should be reviewed thereafter.

**Comment by DL6ER** [comment]
This PR is now ready.

**Comment by yubiuser** [comment]
What's the reason to accept `string` and `[string]` and not just `[string]` for the relevant value? It could simply be an array of 1.

**Comment by DL6ER** [comment]
Following the REST principles, you should add/manipulate one entry per call so sending a string feels natural here. We simply extend this natural feeling by also allowing arrays. You are right, an array of size 1 is exactly identical to a naked string, however, to my understanding, rejecting plain strings and demanding them to be encapsulated in an otherwise empty array shell feels odd to me.

File: src/api/docs/content/specs/common.yaml [hunk 1]
```diff
@@ -15,6 +15,47 @@ components:
           type: string
           description: Key indicating the status of the request
           example: "success"
+    lists_processed:
```
Comment (yubiuser) [inline src/api/docs/content/specs/common.yaml:18]: I see what you're doing here, re-using this part for domains/clients/lists/...
However, it' strange to see `example.com` as part of the client example.

Screenshot at 2023-10-25 22-46-18

File: src/api/list.c [hunk 2]
```diff
@@ -173,69 +179,101 @@ static int api_list_write(struct ftl_conn *api,
...
+					allocated_json = true;
+				}
+				else if(cJSON_IsArray(json_name) && cJSON_GetArraySize(json_name) > 0)
+					row.items = json_name;
 				else
 				{
 					return send_json_error(api, 400,
 					                       "bad_request",
-					                       "Invalid request: No valid item \"name\" in payload",
+					                       "Invalid request: No valid \"name\" in payload (must be either string or array)",
```
Comment (yubiuser) [inline src/api/list.c:230]:
```suggestion
					                       "Invalid request: No valid \"group name\" in payload (must be either string or array)",
```

File: src/api/list.c [hunk 3]
```diff
@@ -308,59 +351,115 @@ static int api_list_write(struct ftl_conn *api,
...
+			if(!cJSON_IsString(it))
+			{
+				okay = false;
+				break;
+			}
+
+			// Check every array element for its validity
+			okay = compile_regex(it->valuestring, &regex, &regex_msg);
+
+			// TODO: The regex needs to be freed here
```
Comment (yubiuser) [inline src/api/list.c:367]: Do you want to do that as part of this PR?

Comment (DL6ER) [reply src/api/list.c]: Done

File: src/api/docs/content/specs/common.yaml (hunk 1 above)
Comment (DL6ER) [reply src/api/docs/content/specs/common.yaml]: I split this into individual parts with matching examples and error messages

**Review by yubiuser — APPROVED** [review]

**Pull request merged**